### PR TITLE
AMFRegressor

### PR DESCRIPTION
**Pull request by kenzabenjelloun**
Hi ! I am working on the AMF Regressor. This is what has been done so far, I tried to keep the same structure as the classifier. There is still a bug somewhere that I can't seem to find.. Thank you for your help !

### CODE REVIEW

**Comment by MaxHalford** [comment]
Hey there, great work! I hope you still managed to find time to enjoy your weekend :)

I'm guessing Saulo will be ok reviewing this PR, although I know he's a bit busy at the moment.

One first comment: since we merged Alexandre's PR, we reorganized stuff and create a new `forest` module. Your fork is missing these changes. At the root of the directory, you can run `make rebase` to pull the changes in the `main` branch. You might have some conflicts to solve, which you will have to do manually. Let me know if this isn't clear. This is actually a great way to learn about git conflicts if this is your first time.

**Comment by smastelini** [comment]
Hi @kenzabenjelloun, thanks for the PR!

As @MaxHalford mentioned, there are some conflicts to solve. Once everything is in place, I will start reviewing the code :D

File: river/ensemble/__init__.py [hunk 1]
```diff
@@ -6,8 +6,11 @@
 aggregate predictions from different kinds of models.
 
 """
+<<<<<<< HEAD
```
Comment (MaxHalford) [inline river/ensemble/__init__.py:9]: It seems you did a good job at merging the `main` branch. However, you're still missing a few cases. I would recommend search for all the `HEAD` mentions and fixing them manually :)

Comment (kenzabenjelloun) [reply river/ensemble/__init__.py]: Hi ! Yes, thank you ! I think it's fixed now.

File: river/forest/aggregated_mondrian_forest.py [hunk 2]
```diff
@@ -228,3 +228,113 @@ def predict_proba_one(self, x):
     @property
     def _multiclass(self):
         return True
+
+
+class AMFRegressor(AMFLearner, base.Regressor):
+    """Aggregated Mondrian Forest regressor for online learning. This algorithm
+    is truly online, in the sense that a single pass is performed, and that predictions
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/forest/aggregated_mondrian_forest.py:235]: Please fix the indentation. The main idea is to separate the first sentence from the other ones.

File: river/forest/aggregated_mondrian_forest.py [hunk 3]
```diff
@@ -228,3 +228,113 @@ def predict_proba_one(self, x):
...
+        The number of trees in the forest.
+    step
+        Step-size for the aggregation weights.
+    use_aggregation
+        Controls if aggregation is used in the trees. It is highly recommended to
+        leave it as `True`.
+    split_pure
+        Controls if nodes that contains only sample of the same class should be
+        split ("pure" nodes). Default is `False`, namely pure nodes are not split,
+        but `True` can be sometimes better.
```
Comment (smastelini) [inline river/forest/aggregated_mondrian_forest.py:262]: I am not sure this applies to regression, since there are no classes.

File: river/forest/aggregated_mondrian_forest.py [hunk 4]
```diff
@@ -228,3 +228,113 @@ def predict_proba_one(self, x):
...
+    seed
+        Random seed for reproducibility.
+
+    Note
+    ----
+    All the parameters of ``AMFRegressor`` become **read-only** after the first call
+    to ``partial_fit``.
```
Comment (smastelini) [inline river/forest/aggregated_mondrian_forest.py:269]: I'm also not sure this applies in this case.

File: river/forest/aggregated_mondrian_forest.py [hunk 5]
```diff
@@ -228,3 +228,113 @@ def predict_proba_one(self, x):
...
+
+    References
+    ----------
+    J. Mourtada, S. Gaiffas and E. Scornet, *AMF: Aggregated Mondrian Forests for Online Learning*, arXiv:1906.10529, 2019
```
Comment (smastelini) [inline river/forest/aggregated_mondrian_forest.py:273]:
```suggestion
    [^1]: J. Mourtada, S. Gaiffas and E. Scornet, *AMF: Aggregated Mondrian Forests for Online Learning*, arXiv:1906.10529, 2019
```

File: river/tree/mondrian/mondrian_tree.py [hunk 6]
```diff
@@ -43,4 +43,4 @@ def __init__(
         self.iteration = iteration
 
         # Controls the randomness in the tree
-        self._rng = random.Random(seed)
+        self._rng = random.Random(seed)
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree.py:46]:
```suggestion
        self._rng = random.Random(seed)

```

Comment (smastelini) [reply river/tree/mondrian/mondrian_tree.py:46]: Please, make sure your editor is not removing the last lines from the file :D

File: river/tree/mondrian/mondrian_tree_nodes.py [hunk 7]
```diff
@@ -366,3 +366,150 @@ class MondrianBranchClassifier(MondrianNodeClassifier, MondrianBranch):
...
+        r = self.predict() - sample_value
+        return r * r / 2
+
+    def update_weight(
+        self,
+        sample_value: base.typing.RegTarget,
+        use_aggregation: bool,
+        step: float,
+    ) -> float:
+        """Update the weight of the node given a class and the method used.
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_nodes.py:411]: Please update the description to account for regression.

File: river/tree/mondrian/mondrian_tree_nodes.py [hunk 8]
```diff
@@ -366,3 +366,150 @@ class MondrianBranchClassifier(MondrianNodeClassifier, MondrianBranch):
...
+        step: float,
+        do_update_weight: bool,
+    ):
+        """Update the node when running a downward procedure updating the tree.
+
+        Parameters
+        ----------
+        x
+            Sample to proceed (as a list).
+        sample_value
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_nodes.py:442]: Please update the description and remove references to classification. I also suggest rewording `x_t` to `x`.

File: river/tree/mondrian/mondrian_tree_nodes.py [hunk 9]
```diff
@@ -366,3 +366,150 @@ class MondrianBranchClassifier(MondrianNodeClassifier, MondrianBranch):
...
     def __init__(self, parent, time, depth, feature, threshold, *children):
         super().__init__(parent, time, depth, feature, threshold, *children)
+
+
+class MondrianNodeRegressor(MondrianNode):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.n_samples = 0
+        self.mean = 0.0
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_nodes.py:376]: You can use an instance of `stats.Mean()` instead, to avoid hardcoding the mean update.

File: river/tree/mondrian/mondrian_tree_regressor.py [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+    """Mondrian Tree Regressor.
+
+    Parameters
+    ----------
+    step
+        Step of the tree.
+    use_aggregation
+        Whether to use aggregation weighting techniques or not.
+    split_pure
+        Whether the tree should split pure leafs during training or not.
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_regressor.py:25]: Same here, maybe it needs to be removed. I don't know how to measure when a node is pure in regression. When the variance is zero?

Are you aware of that @kenzabenjelloun?

File: river/tree/mondrian/mondrian_tree_regressor.py [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+            Split time of the node in the Mondrian process.
+        threshold
+            Threshold of acceptance of the node.
+        feature
+            Feature index of the node.
+        is_right_extension
+            Should we extend the tree in the right or left direction.
+        """
+
+        # Let's build a function to handle splitting depending on what side we go into
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_regressor.py:186]: This comment does not apply here

File: river/tree/mondrian/mondrian_tree_regressor.py [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+            # This test is useless ?
+            if isinstance(current, MondrianLeafRegressor):
+                prediction = self._predict(current)
+            else:
+                weight = current.weight
+                log_weight_tree = current.log_weight_tree
+                w = math.exp(weight - log_weight_tree)
+                # Get the predictions of the current node
+                pred_new = self._predict(current)
+                prediction = 0.5 * w * pred_new + (1 - 0.5 * w) * prediction
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_regressor.py:424]: Does it change in the regression case?

File: river/tree/mondrian/mondrian_tree_nodes.py [hunk 13]
```diff
@@ -366,3 +366,150 @@ class MondrianBranchClassifier(MondrianNodeClassifier, MondrianBranch):
...
+    def loss(self, sample_value: base.typing.RegTarget) -> float:
+        """Compute the loss of the node.
+
+        Parameters
+        ----------
+        sample_value
+            A given value.
+        """
+
+        r = self.predict() - sample_value
```
Comment (smastelini) [inline river/tree/mondrian/mondrian_tree_nodes.py:402]: Is this the correct order in the regression case?

**Review by smastelini — CHANGES_REQUESTED** [review]
Hi @kenzabenjelloun, thanks for bringing the AMF Regressor! I can't wait to put this model to the test :)

I have left some comments in your code, as well as some questions for further clarification.

Some time ago, you mentioned there was a bug that affected performance. Were you able to track this down?

File: river/tree/mondrian/mondrian_tree_regressor.py (hunk 10 above)
Comment (kenzabenjelloun) [reply river/tree/mondrian/mondrian_tree_regressor.py:25]: Hi ! Thank you for taking a look at the code. 
I think it can be removed as it is not really used anyways.

File: river/tree/mondrian/mondrian_tree_regressor.py (hunk 12 above)
Comment (kenzabenjelloun) [reply river/tree/mondrian/mondrian_tree_regressor.py:423]: I don't think so. That's how it is done in the onelearn git (https://github.com/onelearn/onelearn/blob/master/onelearn/tree_methods.py)

File: river/tree/mondrian/mondrian_tree_nodes.py (hunk 13 above)
Comment (kenzabenjelloun) [reply river/tree/mondrian/mondrian_tree_nodes.py:402]: I think so ! It is squared anyways so I don't know if it changes much. 

(https://github.com/onelearn/onelearn/blob/master/onelearn/node_methods.py)

File: river/tree/mondrian/mondrian_tree_regressor.py (hunk 12 above)
Comment (smastelini) [reply river/tree/mondrian/mondrian_tree_regressor.py:423]: Ok! Thanks for the clarification :D

**Comment by smastelini** [comment]
Hi @kenzabenjelloun, was there any progress in finding the performance bug?

I just fixed a conflict with the main branch and resolved some comments you already addressed.

**Comment by smastelini** [comment]
Hi @kenzabenjelloun and @MaxHalford. I've changed the target branch of this PR to a new branch. If you agree, I will merge the PR and take on the remaining work.

**Comment by kenzabenjelloun** [comment]
hi @smastelini ! yes, perfect thank you
